File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int low=0,high=0,ans=0;
        for(int x:weights){
            low=Math.max(low,x);
            high+=x;
        }
        while(low<=high){
            int mid=low+(high-low)/2;
            int dayNumber=1;
            int currentWeight=0;
            for(int w:weights){
                if(currentWeight+w>mid){
                    dayNumber++;
                    currentWeight=w;
                }else{
                    currentWeight+=w;
                }
            }if(dayNumber<=days){
                ans=mid;
                high=mid-1;
            }else{
                low=mid+1;
            }
        }
        return ans;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
private:
    int totalDays(vector<int> &weights,int minWeight){
        int d=1;
        int total=0;
        for(int i=0;i<weights.size();i++){
            if(total+weights[i]<=minWeight){
                total+=weights[i];
            }else{
                d++;
                total=weights[i];
            }
        }
        return d;
    }
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int start=0,end=0;
        for(int i=0;i<weights.size();i++){
            start=max(weights[i],start);
            end+=weights[i];
        }
        int ans=0;
        while(start<=end){
            int mid=start+(end-start)/2;
            int d=totalDays(weights,mid);
            if(d<=days){
                end=mid-1;
            }else{
                start=mid+1;
            }
        }
        return start;
    }
};